### PR TITLE
Core: ADT-ize all errors

### PR DESCRIPTION
**Pull request by jmcardon**
This PR aims to do a few things:
- Add stack frames back to user emitted errors (via `VError` propagating frames in the CEK machine, and the `UserRecoverableError` type in the direct style)
- Fix the pretty printing of all of our errors (a longstanding TODO)
- Clean up code related to messy errors and unused error constructors
- Reevaluate what our actual "Invariant Failures" are, and turn those into an ADT as well.

Post-this PR, the repl output becomes:
```
pact>(ff 1)
(interactive):1:43: boom
 1 | (ff 1)
   |                                            ^^^^^^^^^^^^^^^^^^^^^^
  at (m.f)
  at (m.ff 1)
```

PR checklist:

* [x] Test coverage for the proposed changes
* [x] PR description contains example output from repl interaction or a snippet from unit test output

### CODE REVIEW

File: pact/Pact/Core/IR/Eval/CEK.hs [hunk 1]
```diff
@@ -249,7 +249,8 @@ evaluateTerm cont handler env (Conditional c info) = case c of
   CEnforceOne str conds -> do
     case conds of
       [] ->
-        returnCEK cont handler (VError "enforce-one failure" info)
+        -- Note: this will simply be re-thrown within EnforceErrorC, so we don't need anything fancy here
+        returnCEK cont handler (VError [] (UserEnforceError "") info)
```
Comment (rsoeldner) [inline pact/Pact/Core/IR/Eval/CEK.hs:253]: Maybe we should add some context, even if it will be re-thrown, such as:

```suggestion
        returnCEK cont handler (VError [] (UserEnforceError "internal CEnforceOne error") info)
```

File: pact/Pact/Core/IR/Eval/CEK.hs [hunk 2]
```diff
@@ -404,9 +405,7 @@ applyPact i pc ps cont handler cenv nested = useEvalState esDefPactExec >>= \cas
       -- `initPact` ensures that the step is 0,
       -- and there are guaranteed more than 0 steps due to how the parser is written.
       -- `resumePact` does a similar check before calling this function.
-      when (ps ^. psStep >= nSteps) $ failInvariant i "Step not found"
```
Comment (rsoeldner) [inline pact/Pact/Core/IR/Eval/CEK.hs:407]: Why this change?

File: pact/Pact/Core/IR/Eval/CEK.hs [hunk 3]
```diff
@@ -1286,16 +1300,20 @@ applyContToValue (CondC env info frame cont) handler v = do
           [] -> returnCEKValue cont handler (VList (V.fromList (reverse acc')))
       EnforceOneC ->
         if b then returnCEKValue cont handler v
-        else returnCEK cont handler (VError "enforce-one bool check failure" info)
+        else
+        -- Note: this will simply be re-thrown within EnforceErrorC, so we don't need anything fancy here
+        returnCEK cont handler (VError [] (UserEnforceError "") info)
```
Comment (rsoeldner) [inline pact/Pact/Core/IR/Eval/CEK.hs:1305]: Same as above.

File: pact/Pact/Core/IR/Eval/CEK.hs (hunk 2 above)
Comment (jmcardon) [reply pact/Pact/Core/IR/Eval/CEK.hs:407]: Great question. It's because If you notice below, we actually access the steps partially via a lens and throw an error using `maybe`. It's sort of redundant to throw this invariant error twice.

File: pact/Pact/Core/IR/Eval/CEK.hs [hunk 4]
```diff
@@ -1286,16 +1300,20 @@ applyContToValue (CondC env info frame cont) handler v = do
...
         else returnCEKValue cont handler v
       OrQC clo pv ->
         if not b then applyLam clo [VPactValue pv] (EnforceBoolC info cont) handler
         else returnCEKValue cont handler v
       NotQC -> returnCEKValue cont handler (VBool (not b))
+    VPactValue v' -> throwExecutionError info (ExpectedBoolValue v')
     _ ->
-      returnCEK cont handler (VError "Evaluation of conditional expression yielded non-boolean value" info)
+      -- Note: a non-boolean value in these functions is non recoverable
+      throwExecutionError info ExpectedPactValue
```
[the remark this answers is not in the document]
Comment (jmcardon) [reply pact/Pact/Core/IR/Eval/CEK.hs:1316]: Do you mean the actual value received?

Unfortunately, the problem with this is that closure representations aren't meant to be stable. `TableValue` maybe.

File: pact/Pact/Core/DefPacts/Types.hs [hunk 5]
```diff
@@ -79,3 +80,7 @@ instance NFData Yield
 instance NFData DefPactStep
 instance (NFData name, NFData v) => NFData (DefPactContinuation name v)
 instance NFData DefPactExec
+
+instance (Pretty name, Pretty v) => Pretty (DefPactContinuation name v) where
+  pretty (DefPactContinuation n v) =
+    parens (pretty n <+> if null v then mempty else space <> hsep (pretty <$> v))
```
[the remark this answers is not in the document]
Comment (jmcardon) [reply pact/Pact/Core/DefPacts/Types.hs:90]: Good catch. It actually should be
```
parens (pretty n <> if null v then mempty else space <> hsep (pretty <$> v))
```

[the remark this answers is not in the document]
Comment (jmcardon) [reply pact/Pact/Core/DefPacts/Types.hs:90]: Actually, I've replaced all of these with `PrettyLispApp`

File: pact/Pact/Core/IR/Eval/CoreBuiltin.hs [hunk 6]
```diff
@@ -907,11 +912,12 @@ coreReadInteger info b cont handler _env = \case
             chargeGasArgs info $ GStrOp $ StrOpConvToInt $ T.length raw
             case parseNumLiteral raw of
               Just (LInteger i) -> returnCEKValue cont handler (VInteger i)
-              _ -> returnCEK cont handler (VError "read-integer failure" info)
-
-          _ -> returnCEK cont handler (VError "read-integer failure" info)
-      _ -> returnCEK cont handler (VError "read-integer failure" info)
+              _ -> readError
```
[the remark this answers is not in the document]
Comment (jmcardon) [reply pact/Pact/Core/IR/Eval/CoreBuiltin.hs:915]: Yeah, it looks a bit silly. It's a result of a refactor, I made that `EnvReadFunctionFailure` but it used to take strings and not a `NativeName`. I'll factor it out.

File: pact/Pact/Core/IR/Eval/CoreBuiltin.hs [hunk 7]
```diff
@@ -921,12 +927,14 @@ coreReadMsg info b cont handler _env = \case
         chargeGasArgs info $ GObjOp $ ObjOpLookup s $ M.size envData
         case M.lookup (Field s) envData of
           Just pv -> returnCEKValue cont handler (VPactValue pv)
-          _ -> returnCEK cont handler (VError "read-msg failure" info)
-      _ -> returnCEK cont handler (VError "read-msg failure: data is not an object" info)
+          _ -> readError
```
[the remark this answers is not in the document]
Comment (jmcardon) [reply pact/Pact/Core/IR/Eval/CoreBuiltin.hs]: Done 👍

File: pact/Pact/Core/IR/Eval/Direct/ReplBuiltin.hs [hunk 8]
```diff
@@ -354,18 +354,16 @@ sigKeyset info b _env = \case
...
-          Unmanaged -> do
-            evalCap info env origToken PopCapInvoke TestCapEval cBody
-          _ -> do
-            -- Installed caps emit and event
-            -- so we create a fake stack frame
-            installCap info env origToken False *> evalCap info env origToken PopCapInvoke TestCapEval cBody
-      _ -> throwRecoverableError info "no such capability"
+    d <- getDefCap info (_ctName origToken)
+    let cBody = Constant LUnit info
+        -- cont' = SeqC env cBody cont
```
[the remark this answers is not in the document]
Comment (jmcardon) [reply pact/Pact/Core/IR/Eval/Direct/ReplBuiltin.hs]: Fixed 👍

File: pact-tests/Pact/Core/Test/StaticErrorTests.hs [hunk 9]
```diff
@@ -32,8 +32,11 @@ isDesugarError p s = isJust $ preview (_PEDesugarError . _1 . p) s
 isExecutionError :: Prism' EvalError a -> PactErrorI -> Bool
 isExecutionError p s = isJust $ preview (_PEExecutionError . _1 . p) s
 
-runStaticTest :: String -> Text -> (PactErrorI -> Bool) -> Assertion
-runStaticTest label src predicate = do
+isUserRecoverableError :: Prism' UserRecoverableError a -> PactErrorI -> Bool
+isUserRecoverableError p s = isJust $ preview (_PEUserRecoverableError . _1 . p) s
```
Comment (rsoeldner) [inline pact-tests/Pact/Core/Test/StaticErrorTests.hs:36]:
```suggestion
isUserRecoverableError p = has (_PEUserRecoverableError . _1 . p)
```

**Review by rsoeldner — APPROVED** [review]

File: pact/Pact/Core/Errors.hs [hunk 10]
```diff
@@ -236,6 +263,80 @@ instance Pretty ArgTypeError where
...
+  | InvariantPactExecNotInEnv (Maybe (DefPactContinuation QualifiedName PactValue))
+  -- ^ Defpact Exec expected to be in environment but not found
+  | InvariantPactStepNotInEnv (Maybe (DefPactContinuation QualifiedName PactValue))
+  -- ^ Defpact Exec expected to be in environment but not found
+  | InvariantInvalidManagedCapIndex Int FullyQualifiedName
+  -- ^ managed cap index outside of allowable range
+  | InvariantArgLengthMismatch FullyQualifiedName Int Int
+  -- ^ Argument length mismatch within some internal function
+  | InvariantInvalidManagedCapKind Text
+  -- ^ Invariant managed cap kind, expected, got
```
Comment (0xd34df00d) [inline pact/Pact/Core/Errors.hs:296]: The comment looks trimmed.

File: pact/Pact/Core/Errors.hs [hunk 11]
```diff
@@ -451,40 +604,93 @@ instance Pretty EvalError where
...
+      "SPV verification failure:" <+> pretty e
+    ExpectedBoolValue pv ->
+      "expected bool value, got" <+> pretty pv
+    UserGuardMustBeADefun qn dk ->
+      "User guard closure" <+> pretty qn <+> "must be defun, got" <> pretty dk
+    WriteValueDidNotMatchSchema (Schema _ sc) od ->
+      "Attempted insert failed due to schema mismatch. Expected:" <+> pretty (ObjectData sc)
+        <> ", received" <+> pretty od
+    ObjectIsMissingField f b ->
+      "Key" <+> dquotes (pretty f) <+> "not found in object: " <> pretty b
```
Comment (0xd34df00d) [inline pact/Pact/Core/Errors.hs:669]:
```suggestion
      "Key" <+> dquotes (pretty f) <+> "not found in object:" <> pretty b
```

File: pact/Pact/Core/Errors.hs [hunk 12]
```diff
@@ -451,40 +604,93 @@ instance Pretty EvalError where
...
+    NativeExecutionError n msg ->
+      "native execution failure," <+> pretty n <+> "failed with message:" <+> pretty msg
+    ExpectedStringValue pv ->
+      "expected string value, got: " <+> pretty pv
```
Comment (0xd34df00d) [inline pact/Pact/Core/Errors.hs:673]:
```suggestion
      "expected string value, got:" <+> pretty pv
```

File: pact/Pact/Core/Errors.hs [hunk 13]
```diff
@@ -451,40 +604,93 @@ instance Pretty EvalError where
...
+    ExpectedCapToken pv ->
+      "expected capability token value, got: " <+> pretty pv
```
Comment (0xd34df00d) [inline pact/Pact/Core/Errors.hs:675]:
```suggestion
      "expected capability token value, got:" <+> pretty pv
```

File: pact/Pact/Core/IR/Eval/CEK.hs [hunk 14]
```diff
@@ -1547,28 +1570,30 @@ applyContToValue (ObjC env info currfield fs vs cont) handler v = do
...
+  VPactValue v' -> throwExecutionError info $ ExpectedStringValue v'
+  _ -> throwExecutionError info $ ExpectedPactValue
 -- Discard the value of running a user guard, no error occured, so
 applyContToValue (IgnoreValueC v cont) handler _v =
   returnCEKValue cont handler (VPactValue v)
 
 applyContToValue (StackPopC i mty cont) handler v = do
-  v' <- (\pv -> maybeTCType i pv mty) =<< enforcePactValue i v
+  v' <- enforcePactValue i v
+  maybeTCType i mty v'
```
Comment (0xd34df00d) [inline pact/Pact/Core/IR/Eval/CEK.hs:1583]: Should this be moved out to a separate function to avoid having to call `maybeTCType` (which can be forgotten)?

File: pact/Pact/Core/IR/Eval/Runtime/Utils.hs [hunk 15]
```diff
@@ -346,6 +350,7 @@ enforceStackTopIsDefcap
   -> m ()
 enforceStackTopIsDefcap info b = do
   let (NativeName n) = builtinName b
+  -- Todo: Maybe its own error?
```
Comment (0xd34df00d) [inline pact/Pact/Core/IR/Eval/Runtime/Utils.hs:353]: Leftover or good to go?

File: pact/Pact/Core/Repl/Compile.hs [hunk 16]
```diff
@@ -17,6 +17,8 @@ module Pact.Core.Repl.Compile
  , interpretEvalBigStep
  , interpretEvalSmallStep
  , interpretEvalDirect
+ , interpretReplProgram'
```
Comment (imalsogreg) [inline pact/Pact/Core/Repl/Compile.hs:20]: nit: Slight preference against `'` names. Renaming `interpretReplProgram'` to `interpretReplProgramWithBuiltinEnv`, or renaming `interpretReplProgram` to `interpretReplProgramDefaultBuiltins` is slightly clearer.

File: pact/Pact/Core/Repl/Compile.hs [hunk 17]
```diff
@@ -203,7 +205,7 @@ interpretReplProgram' interpreter (SourceCode _ source) display = do
                 docs <- uses replUserDocs (M.lookup qn)
                 displayValue (RUserDoc d docs)
               Nothing ->
-                failInvariant varI "repl invariant violated: resolved to a top level free variable without a binder"
+                throwExecutionError varI $ EvalError "repl invariant violated: resolved to a top level free variable without a binder"
```
Comment (imalsogreg) [inline pact/Pact/Core/Repl/Compile.hs:208]: ADT-ize this error, too? Or is that going to far?

**Review by 0xd34df00d — APPROVED** [review]
Fantastic work! I left a few very nitpicky comments, please feel free to ignore any!

File: pact/Pact/Core/Type.hs [hunk 18]
```diff
@@ -227,6 +227,16 @@ data DefKind
 
 instance NFData DefKind
 
+instance Pretty DefKind where
+  pretty = \case
+    DKDefun -> "defun"
+    DKDefConst -> "defconst"
+    DKDefCap -> "defcap"
+    DKDefPact -> "defpact"
+    DKDefSchema _ -> "defscema"
```
[the remark this answers is not in the document]
Comment (rsoeldner) [reply pact/Pact/Core/Type.hs:236]: Really good catch 👍  I've overseen it

File: pact-lsp/Pact/Core/LanguageServer.hs [hunk 19]
```diff
@@ -254,7 +254,7 @@ sendDiagnostics nuri mv content = liftIO runPact >>= \case
       PEParseError{} -> "Parse"
       PEDesugarError{} -> "Desugar"
       PEExecutionError{} -> "Execution"
-      PERecoverableError{} -> "Execution"
+      PEUserRecoverableError{} -> "Execution"
```
[the remark this answers is not in the document]
Comment (jmcardon) [reply pact-lsp/Pact/Core/LanguageServer.hs:257]: Great question, @rsoeldner does it matter? I assumed so, it's essentially an error that happens during execution.

File: pact/Pact/Core/StableEncoding.hs [hunk 20]
```diff
@@ -131,7 +131,7 @@ instance J.Encode (StableEncoding KeySetName) where
 instance J.Encode (StableEncoding KeySet) where
   build (StableEncoding (KeySet keys predFun)) =J.object
     [ "pred" J..= StableEncoding predFun
-    , "keys" J..= J.Array (Set.map StableEncoding keys) -- TODO: is this valid?
+    , "keys" J..= J.Array (S.map StableEncoding keys) -- TODO: is this valid?
```
[the remark this answers is not in the document]
Comment (jmcardon) [reply pact/Pact/Core/StableEncoding.hs:134]: I Don't quite remember lol.

File: pact-tests/Pact/Core/Test/StaticErrorTests.hs [hunk 21]
```diff
@@ -573,7 +576,7 @@ executionTests =
         (defun foo:string ())
         )
     |])
-  , ("enforce_ns_define_namespace", isExecutionError _DefineNamespaceError, [text|
+  , ("enforce_ns_define_namespace", isExecutionError _NativeExecutionError, [text|
```
[the remark this answers is not in the document]
Comment (jmcardon) [reply pact-tests/Pact/Core/Test/StaticErrorTests.hs:579]: It might seem like it did, but `NativeExecutionError` includes the native name where it occurred, along with the error message. 

The constructor this replaced was `DefineNamespaceError Text` which would just say something like `Error defining namespace` as opposed to `Error during native execution define-namespace`.

I didn't see a reason for `DefineNamespaceError` when it's simply native + static error message string. Most error constructors that have their own special case, are because we want to hold data to render in them later. In the case of most static error strings, they became `NativeExecutionError`

File: pact/Pact/Core/Errors.hs (hunk 11 above)
Comment (0xd34df00d) [reply pact/Pact/Core/Errors.hs:669]: Dang I missed it's `<>` and not `<+>` lol!

File: pact/Pact/Core/Repl/Compile.hs (hunk 16 above)
Comment (jmcardon) [reply pact/Pact/Core/Repl/Compile.hs:20]: I mean the thing is that it takes an `Interpreter` now, so probably should just be `interpretReplProgram`. The `'` version should be purely internal to the module.

File: pact-tests/Pact/Core/Test/StaticErrorTests.hs (hunk 21 above)
[the remark this answers is not in the document]
Comment (0xd34df00d) [reply pact-tests/Pact/Core/Test/StaticErrorTests.hs:579]: That sounds very reasonable. Agree with this!

File: pact/Pact/Core/Errors.hs (hunk 10 above)
Comment (jmcardon) [reply pact/Pact/Core/Errors.hs:296]: fixed!

File: pact/Pact/Core/IR/Eval/CEK.hs (hunk 14 above)
Comment (jmcardon) [reply pact/Pact/Core/IR/Eval/CEK.hs:1583]: I made an asana ticket for this

File: pact/Pact/Core/Repl/Compile.hs (hunk 16 above)
Comment (jmcardon) [reply pact/Pact/Core/Repl/Compile.hs]: Fixed 👍
